packages/core/src/mixins/closable.js

```diff
@@ -0,0 +1,136 @@
+import React, { PureComponent } from 'react';
+import PropTypes from 'prop-types';
+import keycode from 'keycode';
+
+import getComponentName from '../utils/getComponentName';
+
+const ESCAPE = 'Escape';
+const TOUCH_CLOSE_DELAY = 200;
+
+/**
+ * `closable()` HOC mixin
+ * ====================
+ * Provide various of ways to determine when to “close” a component by
+ * binding event listeners on `document`.
+ *
+ * Formally `escapable()`.
+ *
+ * @param {object} options
+ */
+const closable = ({
+    onEscape = true,
+    onClickOutside = false,
+    onClickInside = false,
+} = {}) => (WrappedComponent) => {
+    const componentName = getComponentName(WrappedComponent);
+
+    class Closable extends PureComponent {
+        static displayName = `closable(${componentName})`;
+
+        static propTypes = {
+            onClose: PropTypes.func,
+        };
+
+        static defaultProps = {
+            onClose: () => {},
+        };
+
+        state = {
+            closeDelayTimeout: null,
+        };
+
+        componentDidMount() {
+            document.addEventListener('keyup', this.handleDocumentKeyup);
+            document.addEventListener('click', this.handleDocumentClick);
+            document.addEventListener('touchstart', this.handleDocumentTouch);
+        }
+
+        componentWillUnmount() {
+            document.removeEventListener('keyup', this.handleDocumentKeyup);
+            document.removeEventListener('click', this.handleDocumentClick);
+            document.removeEventListener('touchstart', this.handleDocumentTouch);
+            clearTimeout(this.state.closeDelayTimeout);
+        }
+
+        // eslint-disable-next-line class-methods-use-this
+        getOptions() {
+            return {
+                onEscape,
+                onClickOutside,
+                onClickInside,
+            };
+        }
+
+        /**
+         * Touchstart event fires as soon as finger touches the screen,
+         * emulate ghost-click's time delay to fire child events first
+         * before trigger the onClose event.
+         *
+         * @ref http://stackoverflow.com/a/9634715
+         */
+        delayedClose = (event) => {
+            const timeout = setTimeout(
```

Comment:
Should clear the timeout when the component was unmount.

If the component was unmounted during the timeout, `this.props` will become undefined and a TypeError will be raised up.

Reply:
I did, at line 52 XD

Reply:
OK, i missed that you set the timeout into the state  😬